### ISSUE TITLE
Discussion - ResponseGenerators should not build `F[Response[F]]`

### ISSUE THREAD

**Issue by lukestephenson** [http4s/http4s#2268]
Currently the [ResponseGenerators](https://github.com/http4s/http4s/blob/9e05425f08ff597553fbfcc3fb75f15aaea0181c/dsl/src/main/scala/org/http4s/dsl/impl/ResponseGenerator.scala#L37) generate `F[Response[F]]`.  For example:
```
val response: IO[Response[IO]] = Ok("hello")
```

Now if I introduce some database access, I might have something like:
```
val userFromDb: IO[User]
val response: IO[Response[IO]] = userFromDb.flatMap { user =>
   Ok("hello ${user}")
}
```

I think this code isn't ideal as there is no `IO` involved in defining the response (streaming it maybe, not objecting to that).   I think the code should ideally be (all I've changed is a `flatMap` to a `map`):
```
val userFromDb: IO[User]
val response: IO[Response[IO]] = userFromDb.map { user =>
   Ok("hello ${user}")
}
```

This is a simplistic example.  If I wanted to introduce the Free monad, then currently I'd end up with:
```
type Script[A] = Free[Actions, A]
val userFromDb: Script[User]
val response: Script[Response[Script]] = userFromDb.flatMap { user =>
   Ok("hello ${user}")
}
```

However `Script[Response[Script]]` is not what I want.  I'd like to build a `Script[Response[IO]]`.  At the controller layer I'm happy for the inner `IO` to be present, but I'd still like to represent the overall application with `Free`.

In summary, I wanted to discuss changing the `ResponseGenerator` to build a `Response[F]`, not a `F[Response[F]]`.

**Comment by rossabaker** [comment]
Through http4s-0.18, encoding a body was an effect: An `EntityEncoder` returned `F[Entity]` rather than `Entity` , so we had no choice.  Now we do.

If we were to make this change, it would have a big effect on simpler routes:

```scala
case GET -> "health-check" =>
  Ok("I'm alive")
``` 

Becomes

```scala
case GET -> "health-check" =>
  Ok("I'm alive").pure[F]
```

That response generator DSL has been effectful for years, so this would be a _massive_ breaking change.

More thoughts on a free monad later, but I'm late for a meeting.  If I haven't responded in 24 hours, give me a nudge.  Also curious what others think.

**Comment by wjlow** [comment]
To add to Luke's comment about Free monad, you end up with a `Script[Response[Script]]` and you're unable to write a natural transformation from that to `IO[Response[IO]]`.

I do understand that this would have a big effect on simple routes though!

**Comment by rossabaker** [comment]
This still works, right?

```scala
userFromDb.map { user => Response[IO](Ok).withEntity(s"hello $user") }
```

i.e., you can do it with the functions on `Message`, just not the syntax from `ResponseGenerator`?

**Comment by lukestephenson** [comment]
Actually the above didn't work.  `Message.withBody` also wraps the result in `IO`.
```
  def withBody[T](b: T)(implicit F: Monad[F], w: EntityEncoder[F, T]): F[Self] =
    w.toEntity(b).flatMap { entity =>
      val hs = entity.length match {
        case Some(l) =>
          `Content-Length`
            .fromLong(l)
            .fold(
              _ =>
                F.pure {
                  Message.logger.warn(s"Attempt to provide a negative content length of $l")
                  w.headers.toList
              },
              cl => F.pure(cl :: w.headers.toList))
        case None => F.pure(w.headers)
      }
      hs.map(newHeaders => change(body = entity.body, headers = headers ++ newHeaders))
    }
```

Was `withEntity` a typo, or does that behave differently to `withBody`?

**Comment by rossabaker** [comment]
`withBody` was deprecated in favor of `withEntity` when `EntityEncoder` became pure instead of effectful:

```scala
  @deprecated("Use withEntity", "0.19")
  def withBody[T](b: T)(implicit F: Applicative[F], w: EntityEncoder[F, T]): F[Self]

  def withEntity[T](b: T)(implicit w: EntityEncoder[F, T]): Self
```

**Comment by lukestephenson** [comment]
Ok.  Still on `0.18` so will have to upgrade at some point.

Thanks again for all the great work and responding to this.
